### DISCUSSION TITLE
Mapping Messages in the `update` function

### DISCUSSION THREAD

**Discussion by andretietz** [hecrj/iced#1899] [Q&A]
When I implement the `update` function of `Application` sometimes I would like to see a mapping of my message.

I know this seems odd in the first place, since I could directly send the message I want a command to follow on, but have a look on the following example:

```rust

ScreenMessage::Running(running) => {
        return match running {
          RunningMessage::TabSelected(selected) => {
            let init_tab = self.active_tab != selected;
            self.active_tab = selected;
            if init_tab {
              return match self.active_tab {
                TabId::Service => {
                  let message = self.service_tab.on_select();
                  Command::perform(async {}, |_| ScreenMessage::Running(RunningMessage::Service(message)))
                }
                TabId::Customers => {
                  let message = self.customers_tab.on_select();
                  Command::perform(async {}, |_| ScreenMessage::Running(RunningMessage::Customers(message)))
                }
                TabId::Invoices => {
                  let message = self.invoices_tab.on_select();
                  Command::perform(async {}, |_| ScreenMessage::Running(RunningMessage::Invoices(message)))
                }
                TabId::Settings => {
                  let message = self.settings_tab.on_select();
                  Command::perform(async {}, |_| ScreenMessage::Running(RunningMessage::Settings(message)))
                }
              };
            }
            Command::none()
          }
          RunningMessage::Customers(message) => self
            .customers_tab
            .update(message)
            .map(|m| ScreenMessage::Running(RunningMessage::Customers(m))),
          RunningMessage::Service(message) => self
            .service_tab
            .update(message)
            .map(|m| ScreenMessage::Running(RunningMessage::Service(m))),
          RunningMessage::Invoices(message) => self
            .invoices_tab
            .update(message)
            .map(|m| ScreenMessage::Running(RunningMessage::Invoices(m))),
          RunningMessage::Settings(message) => self
            .settings_tab
            .update(message)
            .map(|m| ScreenMessage::Running(RunningMessage::Settings(m))),
        };
      }

```

Whenever my tab gets selected, I want to reinitialize it, this is why I am calling `on_select()` and afterwards there's nothing to be done, but to forward the message into the tab. I am using e.g. 
```rust 
Command::perform(async {}, |_| ScreenMessage::Running(RunningMessage::Settings(message))) 
``` 
in order to do this "mapping".

Now one might say "well why don't you "map" this directly in the callback of tabselection instead of the `update()` method itself?"

1. I am not sure if `update()` and `view()` (where I set the callback) are running in the same Thread/Coroutine. If they do, this is likely subject of change. I want to keep logic out of the `view()` method.
2. Calling the `on_select` method within a callback requires Boxing, which seems unnessessary for this case.

May be I am already able to do this kind of mapping, without creating an empty async function, if so, I don't know how.

**Comment by lufte** [comment]
I think this is similar to the discussion at #1172. You basically have all the information you need to make your decision when the first message arrives, so you should be able to structure your code in a way that allows you to run all the required logic at that point without resorting to manually dispatching messages nor duplicating code.
